File: day_3/day_3.dart
// https://adventofcode.com/2021/day/3

import 'dart:io';

enum CommonType {
  most,
  least,
}

// Calculates a new binary value based on the most common value at each index
String calculateCommon(
  List<String> values, {
  required CommonType commonType,
}) {
  String common = '';

  // Split values into lists of individual binary values
  List<List<String>> splitLists = [];

  values.forEach((value) {
    splitLists.add(value.trim().split(''));
  });

  // Group binary values by index
  List<String> groupedBitsStrings = [];

  splitLists.forEach((list) {
    for (var i = 0; i < list.length; i++) {
      // Add an empty string for index if missing
      if (groupedBitsStrings.length < i + 1) {
        groupedBitsStrings.add('');
      }

      groupedBitsStrings[i] += list[i];
    }
  });

  // Calculate the most common bits for each index and add to common string
  groupedBitsStrings.forEach((indexGroup) {
    int zeroCount = '0'.allMatches(indexGroup).length;
    int oneCount = '1'.allMatches(indexGroup).length;

    switch (commonType) {
      case (CommonType.most):
        if (zeroCount > oneCount) {
          common += '0';
        } else {
          common += '1';
        }
        break;
      case (CommonType.least):
        if (zeroCount <= oneCount) {
          common += '0';
        } else {
          common += '1';
        }
        break;
    }
  });

  return common;
}

main() async {
  final File file = File(Directory.current.path + '/day_3/input.txt');
  List<String> values = await file.readAsLines();

  //! PART 1
  String mostCommon = calculateCommon(values, commonType: CommonType.most);
  String leastCommon = calculateCommon(values, commonType: CommonType.least);

  // Convert to decimal from binary
  int gamma = int.parse(mostCommon, radix: 2);
  int epsilon = int.parse(leastCommon, radix: 2);

  print(gamma * epsilon);

  //! PART 2
  // Create copies of values for oxygen and c02
  List<String> oxygenValues = [...values];
  List<String> c02Values = [...values];

  // Loop for each index
  for (var i = 0; i < values[0].length; i++) {
    // Recalculate common result based on current oxygen and c02 values
    String currentMostCommon = calculateCommon(
      oxygenValues,
      commonType: CommonType.most,
    );
    String currentLeastCommon = calculateCommon(
      c02Values,
      commonType: CommonType.least,
    );

    // Remove values that to not match the current index
    if (oxygenValues.length > 1) {
      oxygenValues.removeWhere((value) => value[i] != currentMostCommon[i]);
    }
    if (c02Values.length > 1) {
      c02Values.removeWhere((value) => value[i] != currentLeastCommon[i]);
    }
  }

  // Convert to decimal from binary
  final int oxygenGenRating = int.parse(oxygenValues[0], radix: 2);
  final int c02GenRating = int.parse(c02Values[0], radix: 2);

  print(oxygenGenRating * c02GenRating);
}

File: day_3/day_3.py
import copy
import os

def calculate_most_common(values):
    bit_lists = []
    for value in values:
        bit_lists.append([char for char in value])

    grouped_bits_list = []
    for bits in bit_lists[0]:
        grouped_bits_list.append([])

    i = 0
    while i < 12:
        for list in bit_lists:
            for index, bit in enumerate(list):
                grouped_bits_list[index].append(bit)
        i += 1
    
    most_common = ''

    for group in grouped_bits_list:
        zeros = group.count('0')
        ones = group.count('1')

        if zeros > ones:
            most_common += '0'
        else:
            most_common += '1'
    
    return most_common

def calculate_least_common(values):
    bit_lists = []
    for value in values:
        bit_lists.append([char for char in value])

    grouped_bits_list = []
    for bits in bit_lists[0]:
        grouped_bits_list.append([])

    i = 0
    while i < 12:
        for list in bit_lists:
            for index, bit in enumerate(list):
                grouped_bits_list[index].append(bit)
        i += 1
    
    least_common = ''

    for group in grouped_bits_list:
        zeros = group.count('0')
        ones = group.count('1')

        if zeros <= ones:
            least_common += '0'
        else:
            least_common += '1'
    
    return least_common


def main():
    __location__ = os.path.realpath(os.path.join(os.getcwd(), os.path.dirname(__file__)))
    with open(os.path.join(__location__,'input.txt')) as file:
        values = file.readlines()
        values = [value.rstrip() for value in values]

    # Part 1
    most_common = calculate_most_common(values)
    least_common = calculate_least_common(values)

    gamma_rate = f'0b{most_common}'
    epsilon_rate = f'0b{least_common}'

    print(int(gamma_rate, 2) * int(epsilon_rate, 2))

    # Part 2
    oxygen_values = copy.copy(values)
    c02_values = copy.copy(values)

    i = 0
    while i < 12:
        oxygen_values_to_remove = []
        c02_values_to_remove = []

        most_common = calculate_most_common(oxygen_values)
        least_common = calculate_least_common(c02_values)
        
        # Calculate Oxygen
        for value in oxygen_values:
            if value[i] != most_common[i]:
                oxygen_values_to_remove.append(value)
        
        if len(oxygen_values_to_remove) < len(oxygen_values):
            for value_to_remove in oxygen_values_to_remove:
                oxygen_values.remove(value_to_remove)

        # Calculate C02
        for value in c02_values:
            if value[i] != least_common[i]:
                c02_values_to_remove.append(value)
        
        if len(c02_values_to_remove) < len(c02_values):
            for value_to_remove in c02_values_to_remove:
                c02_values.remove(value_to_remove)

        i += 1
    
    # Convert from binary to int and multiply
    print(int(f'0b{oxygen_values[0]}', 2) * int(f'0b{c02_values[0]}', 2))

if __name__ == "__main__":
    main()